pkg/vm/engine/disttae/types.go

```diff
@@ -17,19 +17,58 @@ package disttae
 import (
 	"sync"
 
+	"github.com/matrixorigin/matrixone/pkg/container/batch"
+	"github.com/matrixorigin/matrixone/pkg/pb/logservice"
+	"github.com/matrixorigin/matrixone/pkg/pb/timestamp"
 	"github.com/matrixorigin/matrixone/pkg/pb/txn"
 )
 
+const (
+	// default database name for catalog
+	MO_CATALOG  = "mo_catalog"
+	MO_DATABASE = "mo_database"
+	MO_TABLES   = "mo_tables"
+	MO_COLUMNS  = "mo_columns"
+)
+
+type Reader interface {
+}
+
 type Engine struct {
-	sync.Mutex
+	sync.RWMutex
 	getClusterDetails GetClusterDetailsFunc
 	txns              map[string]*Transaction
 }
 
+// ReadOnly DB cache for tae
+type DB struct {
+	readTs timestamp.Timestamp
+}
+
 // Transaction represents a transaction
 type Transaction struct {
 	// readOnly default value is true, once a write happen, then set to false
-	readOnly    bool
-	statementId [2]uint64
+	readOnly bool
+	db       *DB
+	// blockId starts at 0 and keeps incrementing,
+	// this is used to name the file on s3 and then give it to tae to use
+	blockId uint64
+	// use for solving halloween problem
+	statementId uint64
 	meta        txn.TxnMeta
+	// fileMaps used to store the mapping relationship between s3 filenames and blockId
+	fileMap map[string]uint64
+	// writes cache stores any writes done by txn
+	// every statement is an element
+	writes   [][]Entry
+	dnStores []logservice.DNStore
+}
+
+// Entry represents a delete/insert
+type Entry struct {
```

Comment:
How can you tell if this entry is an insert of delete?

Reply:
My question, I missed the type. Embarrassment. I'll add it later

Reply:
Where is rowid?   Inside batch?

Reply:
yes, The data for rowid is a column.